### ISSUE TITLE
Lambda: cam Acquire must wait for IMMout Capture to finish

### ISSUE THREAD

**Issue by prjemian** [aps-8id-trr/ipython-8idiuser#74]
The Lambda detector IOC cam plugin turns off `Acquire` *before* the immout plugin `Capture` is finished.  Some frames are not written to output.  Need to set some *block* parameter(s) in the IOC.

**Comment by prjemian** [comment]
note: https://github.com/aps-8id-trr/ipython-8idiuser/blob/b9d28f3107cd693a377a4ad7adec6a4ae0791fdf/user/ad_acquire.py#L192

**Comment by prjemian** [comment]
In AD 3.4, there is a wait-for-plugins feature but we are using AD 3.2.0.  We must override the default ophyd `trigger` method for the adlambda to include the status of the immout:num_captured that it reaches the desired number of images or times out.  Maybe we can just look at the immout.capture value to reach Done.  Both the cam.acquire and the immout.capture must be done before we declare the detector acquisition is complete.  Needs a DeviceStatus object on the immout.capture and some more finesse.

**Comment by prjemian** [comment]
Tried technique from offline system (wait for status of `adlambda.immout.capture.value == 0`) but all plugins stopped when `adlambda.cam.acquire.value == 0`.  There were still unprocessed frames left in the plugin queue.

Tried making a custom IMMOutTriggerStatus() class but still same problem.

Can we simulate with a sim detector and a plugin that is not blocking acquisition?

**Comment by prjemian** [comment]
On the (areaDetector/ADCore) NDPluginBaseFull screen, there is an option for *Callbacks block* which is set to *No* by default.  This corresponds to `<PREFIX>:<PLUGIN>:BlockingCallbacks` and  `<PREFIX>:<PLUGIN>:BlockingCallbacks_RBV`

**Comment by prjemian** [comment]
Here's a view for the `HDF1` plugin with `BlockingCallbacks` set to *Yes*:

![Clipboard01](https://user-images.githubusercontent.com/2279984/66669741-c2ec8d00-ec1d-11e9-90dc-bd4c5b83a35e.jpg)

**Comment by prjemian** [comment]
With `BlockCallbacks` set to *No*, the HDF5 plugin lags a bit after `cam1:Acquire` goes back to 0:

```
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:27:11.871596 60  
adsky:cam1:Acquire             2019-10-11 10:27:11.871778 Done  
open file "/tmp/docker_ioc/synapps-6.1-ad-3.7/screens/ui/NDPluginBaseFull.ui" with macro "P=adsky:,R=HDF1:"
adsky:HDF1:Capture_RBV         2019-10-11 10:28:01.915563 Capturing STATE MINOR
adsky:cam1:Acquire             2019-10-11 10:28:27.959306 Acquire  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:27.970668 61  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:27.978480 61  
adsky:HDF1:QueueFree           2019-10-11 10:28:27.978874 19  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:27.978883 62  
adsky:HDF1:QueueFree           2019-10-11 10:28:27.979770 20  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:27.979866 62  
adsky:HDF1:QueueFree           2019-10-11 10:28:27.980994 19  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:27.981000 63  
adsky:HDF1:QueueFree           2019-10-11 10:28:27.984505 20  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:27.984601 63  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:27.986229 64  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:27.992207 64  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:27.993446 65  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:27.995296 65  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.001976 66  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.006992 66  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.008763 67  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.018353 67  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.018365 19  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.018371 68  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.022098 68  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.022108 69  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.027537 69  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.027553 70  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.036439 70  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.036451 71  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.037648 18  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.037654 72  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.050067 19  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.050728 71  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.051122 18  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.051129 73  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.055982 19  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.056336 72  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.056656 18  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.056663 74  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.062456 19  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.062851 73  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.063295 18  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.063302 75  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.072859 74  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.072872 76  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.073900 17  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.073907 77  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.080602 75  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.080612 78  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.092951 16  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.092960 79  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.094911 17  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.095025 76  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.100678 16  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:28:28.100686 80  
adsky:cam1:Acquire             2019-10-11 10:28:28.100704 Done  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.106937 17  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.107076 77  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.108045 18  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.108130 78  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.111128 19  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.111271 79  
adsky:HDF1:QueueFree           2019-10-11 10:28:28.122650 20  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:28:28.122782 80  
adsky:HDF1:Capture_RBV         2019-10-11 10:28:28.122983 Done  
```

**Comment by prjemian** [comment]
After setting `HDF1:BlockingCallbacks` to *Yes*, then `cam1:Acquire` does not go back to 0 until *after* the HDF1 plugin is done processing its queue:

```
adsky:HDF1:Capture_RBV         2019-10-11 10:29:24.372777 Capturing STATE MINOR
adsky:cam1:Acquire             2019-10-11 10:29:34.029223 Acquire  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.032357 81  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.032375 81  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.039827 82  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.039982 82  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.044546 83  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.044744 83  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.052356 84  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.053843 84  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.060532 85  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.060699 85  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.074073 86  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.074295 86  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.082047 87  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.082840 87  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.090557 88  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.091351 88  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.099368 89  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.099544 89  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.110256 90  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.110428 90  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.115500 91  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.115680 91  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.128528 92  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.128715 92  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.137592 93  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.137808 93  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.143018 94  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.143175 94  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.157649 95  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.157828 95  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.161706 96  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.162642 96  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.170740 97  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.171660 97  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.176386 98  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.178242 98  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.184376 99  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.189705 99  
adsky:HDF1:UniqueId_RBV        2019-10-11 10:29:34.198205 100  
adsky:HDF1:Capture_RBV         2019-10-11 10:29:34.198423 Done  
adsky:cam1:ArrayCounter_RBV    2019-10-11 10:29:34.198566 100  
adsky:cam1:Acquire             2019-10-11 10:29:34.198773 Done  
```

**Comment by prjemian** [comment]
We should examine the setting of `IMMout:BlockingCallbacks`.  Try it with *Yes*.  Does that add unacceptable, random time jitter?

**Comment by prjemian** [comment]
Currently:

```
snow% caget 8LAMBDA1:IMMout:BlockingCallbacks_RBV
8LAMBDA1:IMMout:BlockingCallbacks_RBV No
snow% caget 8LAMBDA1:IMMout:QueueFree
8LAMBDA1:IMMout:QueueFree      30000
```

Changing `8LAMBDA1:IMMout:BlockingCallbacks` to *Yes* is the most natural way to resolve the problem, as long as the time jitter is not compromised.

**Comment by prjemian** [comment]
Alas, we are already setting `8LAMBDA1:IMMout:BlockingCallbacks`: https://github.com/aps-8id-trr/ipython-8idiuser/blob/f6392dbe62db2e9d1f8feca2fac2e916d59a6f71/profile_bluesky/startup/25-areadetector-lambda.py#L215

so that is **not** the problem here.

**Comment by prjemian** [comment]
Continued testing the simulator with bluesky.  When acquiring 1000 frames and writing to HDF5 with a 20-frame buffer and **no blocking**, the HDF plugin drops hundreds of frames.  The cam plugin returns to 0 about 2 seconds before the HDF plugin.  But, the HDF plugin is allowed to finish:

```
adsky:HDF1:DroppedArrays_RBV   2019-10-11 15:49:15.611277 142  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.635643 1 LOLO MAJOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.646599 2 LOW MINOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.647104 1 LOW MINOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.655873 0 LOLO MAJOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.660562 1 LOLO MAJOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.660943 0 LOLO MAJOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.681253 1 LOLO MAJOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.681607 0 LOLO MAJOR
adsky:cam1:Acquire             2019-10-11 15:49:15.683008 Done  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.686682 1 LOLO MAJOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.690854 2 LOW MINOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.695222 3 LOW MINOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.700409 4 LOW MINOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.703306 5 LOW MINOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.707775 6 LOW MINOR
adsky:HDF1:QueueFree           2019-10-11 15:49:15.712333 7  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.715486 8  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.723582 9  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.727735 10  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.732448 11  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.737154 12  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.740771 13  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.745805 14  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.748923 15  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.754620 16  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.759685 17  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.766231 18  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.770091 19  
adsky:HDF1:QueueFree           2019-10-11 15:49:15.771688 20  
adsky:HDF1:Capture_RBV         2019-10-11 15:49:16.508205 Done  
```

With our IMMout plugin, the plugin is turned off when the cam plugin is done, it seems.

**Comment by sureshnaps** [comment]
@prjemian I think the cleanest solution is if there is a hook available in ophyd that we can use. In spec, we will take the NOT of all the finish PVs and AND them, so it is safer. My feeling is the IMM plugin is a bit non-standard and might do things differently. Is such a hook possible?

**Comment by prjemian** [comment]
Probably.  We should work up an easier test case than the AD_Acquire plan, a test case that exercises all the detector features we need without engaging the data management workflow or bothering any other beam line settings.  (Read-only on any 8idi:StrRegNNN or 8idi:RegN PVs.)

**Comment by sureshnaps** [comment]
Good idea. Does this make sense. In the transfer(0)/analysis(1) PV, should we consider a -1 which will not call any DM or any other setting. It will just do the AD stuff.

**Comment by prjemian** [comment]
works for me.

**Comment by prjemian** [comment]
I'll come on Monday morning, 10:00-ish

**Comment by sureshnaps** [comment]
ok thanks. The users will be running Rigaku but we can test lambda. We just need to comment out the pre and post shutter opening and closing if that is involved.

**Comment by prjemian** [comment]
Pausing work on this issue in favor of #84.

**Comment by prjemian** [comment]
Since a **new** EPICS areaDetector driver for this camera is in progress, marked this issue as *wontfix* for now (see #85).  Removing milestone as well. Not closing since we may come back to this point.

**Comment by prjemian** [comment]
Closing.  Re-open or open new as needed.
